Rework JSON Serialization section (#414)
oauth-wg/oauth-selective-disclosure-jwt: draft-ietf-oauth-selective-disclosure-jwt.md

# JWS JSON Serialization {#json_serialization}

This section describes an alternative format for SD-JWT using the JWS JSON
Serialization from [@!RFC7515]. Supporting this format is OPTIONAL.

## New Unprotected Header Parameters {#json_serialization_unprotected_headers}

For both the General and Flattened JSON Serialization, the SD-JWT is represented
as a JSON object according to Section 7.2 of [@!RFC7515]. The following new
unprotected header parameters are defined:

 * `disclosures`: An array of strings where each element is an individual
   Disclosure as described in (#creating_disclosures).
 * `kb_jwt`: A Key Binding JWT as described in (#kb-jwt).

If a Key Binding JWT is present, the digest in the `sd_hash` claim MUST be taken
over a string built as described in (#integrity-protection-of-the-presentation).
The "Issuer-signed JWT" part is built by concatenating the protected header, the
payload, and the signature of the JWS JSON serialized SD-JWT using a `.`
character as a separator, and using the Disclosures from the `disclosures`

> member of the unprotected header. In case of multiple signatures, only the first

Reviewer: disclosures and paylaod need to be base64 url encoded before being concatenated with `.`, no?

Reviewer: ok, I read unprotected header instead of protected header, all good.

> one is used.

Reviewer 2: My first read of this last sentence had me thinking it was only about using the first signature value for constructing the "Issuer-signed JWT" part for input to the `sd_hash`. 

But I think it's also trying to say that, in the case of multiple signatures in the general serialization, the `disclosures` can only show up (and be used for constructing the `sd_hash` input) in the `header` of first of the `signatures` array.  And that the `kb_jwt` has to be there too.

And now I see that is what is said in the General JSON Serialization subsection below. Sorry!

But a bit more clarity here might still be good? I don't know that this is that much better but maybe:

```suggestion
one is used for the Disclosures and Key Binding JWT. 
```

Side note: using only the first one feels kinda icky TBH and the general issue was one reason I'd tried to avoid using the unprotected header for the SD-JWT pieces. But it makes sense to do it this way given all the factors and constraints involved.

Author: Thanks, I think the clarification is useful! If we can come up with a better way than just using the first disclosure, I would be open to that.

## Flattened JSON Serialization

In case of the Flattened JSON Serialization, there is only one unprotected
header.

The following is a non-normative example of a JWS JSON serialized SD-JWT as
issued using the Flattened JSON Serialization:

<{{examples/json_serialization_flattened/sd_jwt_issuance.json}}

The following is a presentation including a Key Binding JWT and two Disclosures:

<{{examples/json_serialization_flattened/sd_jwt_presentation.json}}

## General JSON Serialization

In case of the General JSON Serialization, there are multiple unprotected
headers (one per signature). If present, `disclosures` and `kb_jwt`, MUST be
included in the first unprotected header and MUST NOT be present in any
following unprotected headers.

The following is a non-normative example of a presentation of a JWS JSON
serialized SD-JWT including a Key Binding JWT using the General JSON
Serialization:

<{{examples/json_serialization_general/sd_jwt_presentation.json}}

## Verification of the JWS JSON Serialized SD-JWT

> Verification of the JWS JSON serialized SD-JWT follows the same rules defined in

Author: ```suggestion
Verification of the JWS JSON serialized SD-JWT follows the rules defined in
```

(#verification), except for the following aspects:

 * The SD-JWT does not need to be split into component parts and the Disclosures

>    can be found in the respective member of the JSON object.

Reviewer: Disclosures can be found in the unprotected header, no? I am probably just not reading this right..

Reviewer: ```suggestion
   can be found in the respective member of the JSON object in the unprotected header.
```

>    can be found in the respective member of the JSON object.

Author: ```suggestion
   can be found in the `disclosures` member of the unprotected header.
```
